File: src/utils/IOUtils.java
/*
 * Copyright LWJGL. All rights reserved.
 * License terms: https://www.lwjgl.org/license
 */
package utils;

import jdk.internal.org.xml.sax.SAXException;
import org.lwjgl.*;
import org.w3c.dom.Document;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.*;
import java.nio.*;
import java.nio.channels.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;

import static org.lwjgl.BufferUtils.*;

public class IOUtils {

    private static ByteBuffer resizeBuffer(ByteBuffer buffer, int newCapacity) {
        ByteBuffer newBuffer = BufferUtils.createByteBuffer(newCapacity);
        buffer.flip();
        newBuffer.put(buffer);
        return newBuffer;
    }

    /**
     * Reads the specified resource and returns the raw data as a ByteBuffer.
     *
     * @param resource   the resource to read
     * @param bufferSize the initial buffer size
     *
     * @return the resource data
     *
     * @throws IOException if an IO error occurs
     */
    public static ByteBuffer ioResourceToByteBuffer(String resource, int bufferSize) throws IOException {
        ByteBuffer buffer;

        Path path = Paths.get(resource);
        if (Files.isReadable(path)) {
            try (SeekableByteChannel fc = Files.newByteChannel(path)) {
                buffer = BufferUtils.createByteBuffer((int)fc.size() + 1);
                while (fc.read(buffer) != -1) {
                    ;
                }
            }
        } else {
            try (
                InputStream source = IOUtils.class.getClassLoader().getResourceAsStream(resource);
                ReadableByteChannel rbc = Channels.newChannel(source)
            ) {
                buffer = createByteBuffer(bufferSize);

                while (true) {
                    int bytes = rbc.read(buffer);
                    if (bytes == -1) {
                        break;
                    }
                    if (buffer.remaining() == 0) {
                        buffer = resizeBuffer(buffer, buffer.capacity() * 3 / 2); // 50%
                    }
                }
            }
        }

        buffer.flip();
        return buffer;
    }

    public static Document stringToXmlDocument(String xmlString) {
        try {
            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            InputStream inputStream = new ByteArrayInputStream(xmlString.getBytes(StandardCharsets.UTF_8));
            return builder.parse(inputStream);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static void stringToXmlFile(String xmlSource, String pathFile)
            throws SAXException, ParserConfigurationException, IOException, TransformerException, org.xml.sax.SAXException {
        // Parse the given input
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        InputStream inputStream = new ByteArrayInputStream(xmlSource.getBytes(StandardCharsets.UTF_8));
        Document doc = builder.parse(inputStream);

        // Write the parsed document to an xml file
        TransformerFactory transformerFactory = TransformerFactory.newInstance();
        Transformer transformer = transformerFactory.newTransformer();
        DOMSource source = new DOMSource(doc);

        StreamResult result =  new StreamResult(new File(pathFile));
        transformer.transform(source, result);
    }
}

File: src/utils/IOUtils.kt
package utils;

import java.io.*
import java.nio.*
import java.nio.channels.*
import java.nio.charset.Charset
import java.nio.file.*
import org.lwjgl.BufferUtils.*

/**
 * Reads the specified resource and returns the raw data as a ByteBuffer.
 *
 * @param resource   the resource to read
 * @param bufferSize the initial buffer size
 *
 * @return the resource data
 *
 * @throws IOException if an IO error occurs
 */
fun ioResourceToByteBuffer(resource: String, bufferSize: Int): ByteBuffer {
  val path = Paths.get(resource)
  if (Files.isReadable(path)) {
    val fc = Files.newByteChannel(path)
    val buffer = createByteBuffer(fc.size().toInt() + 1)
    while (fc.read(buffer) != -1) {
      //Nothing
    }
    buffer.flip()
    return buffer
  } else {
    val buffer = createByteBuffer(bufferSize);
    object {}.javaClass.classLoader.getResourceAsStream(resource)?.let { inputStream ->
      val rbc = Channels.newChannel(inputStream)
      while (true) {
        val bytes = rbc.read(buffer);
        if (bytes == -1) {
          break;
        }
        if (buffer.remaining() == 0) {
          resizeBuffer(buffer, buffer.capacity() * 3 / 2); // 50%
        }
      }
      buffer.flip()
    }
    return buffer
  }
}

fun readFile(path: String, encoding: Charset): String {
  val encoded = Files.readAllBytes(Paths.get(path))
  return String(encoded, encoding)
}

private fun resizeBuffer(buffer: ByteBuffer, newCapacity: Int): ByteBuffer {
  val newBuffer = createByteBuffer(newCapacity)
  buffer.flip()
  newBuffer.put(buffer)
  return newBuffer
}
